medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,590 @@
+from medmodels.medrecord.querying import edge, node
+from medmodels.medrecord.types import (
+    Group,
+    MedRecordAttribute,
+    MedRecordValue,
+    NodeIndex,
+)
+from medmodels.treatment_effect.temporal_analysis import find_reference_edge
+
+
+class PreprocessingHyperparameters(TypedDict, total=True):
+    minimum_occurrences_concept: int
+    time_interval_days: int
+    minimum_concepts_per_window: int
+    number_of_sampled_patients: int
+
+
+class PreprocessingHyperparametersOptional(TypedDict, total=False):
+    minimum_occurrences_concept: int
+    time_interval_days: int
+    minimum_concepts_per_window: int
+    number_of_sampled_patients: int
+
+
+class PreprocessingAttributes(TypedDict):
+    first_admission_attribute: MedRecordAttribute
+    time_window_attribute: MedRecordAttribute
+    concept_index_attribute: MedRecordAttribute
+    number_of_windows_attribute: MedRecordAttribute
+    absolute_time_window_attribute: MedRecordAttribute
+
+
+class MTGANPreprocessor(nn.Module):
+    """Preprocessor for the MTGAN model."""
+
+    patients_group: Group
+    concepts_group: Group
+    time_attribute: MedRecordAttribute
+    hyperparameters: PreprocessingHyperparameters
+    seed: int
+
+    def __init__(
+        self,
+        hyperparameters: PreprocessingHyperparameters,
+        patients_group: Group = "patients",
+        concepts_group: Group = "concepts",
+        time_attribute: MedRecordAttribute = "time",
+        seed: int = 42,
+    ) -> None:
+        """Initialize the MTGANPreprocessor.
+
+        Args:
+            hyperparameters (PreprocessingHyperparameters): The hyperparameters for the
+                preprocessing.
+            patients_group (Group): The group of patients in the MedRecord.
+            concepts_group (Group): The group of concepts in the MedRecord.
+            time_attribute (MedRecordAttribute): The attribute name for the time in the
+                edges of the MedRecord.
+            seed (int): The seed for the random number generator. Default is 42.
+        """
+        super().__init__()
+        self.patients_group = patients_group
+        self.concepts_group = concepts_group
+        self.time_attribute = time_attribute
+        self.hyperparameters = hyperparameters
+        self.seed = seed
+
+    def _get_attribute_name(
+        self, medrecord: MedRecord, attribute_name: MedRecordAttribute
+    ) -> MedRecordAttribute:
+        """Generates a unique attribute name.
+
+        If the attribute name is "time", it will return "time_1" if "time" already
+        exists in the MedRecord. If "time_1" exists, it will return "time_2", and so
+        on.
+
+        Args:
+            medrecord (MedRecord): The MedRecord.
+            attribute_name (MedRecordAttribute): The attribute name.
+
+        Returns:
+            MedRecordAttribute: Unique attribute name for the MedRecord.
+        """
+        created_name, counter = attribute_name, 1
```

Comment:
should not be used except when functions return multiple values

```suggestion
        created_name = attribute_name,
        counter = 1
```